src/plonk_recursion.rs

```diff
@@ -170,67 +220,147 @@ fn verify_all_ipas<C: HaloCurve, InnerC: HaloCurve<BaseField=C::ScalarField>>(
         proof.c_wires.clone(),
         vec![proof.c_plonk_z],
         proof.c_plonk_t.clone(),
-    ].concat();
+    ]
+    .concat();
     let mut c_reduction_muls = Vec::new();
     let powers_of_u = powers_recursive(builder, u, c_all.len());
     for (&c, &power) in c_all.iter().zip(powers_of_u.iter()) {
-        let mul = CurveMulOp { scalar: power, point: c };
+        let mul = CurveMulOp {
+            scalar: power,
+            point: c,
+        };
         c_reduction_muls.push(mul);
     }
     let c_reduction_msm_result = builder.curve_msm_endo::<InnerC>(&c_reduction_muls);
     let actual_scalars = c_reduction_msm_result.actual_scalars;
     let c_reduction = c_reduction_msm_result.msm_result;
 
     // For each opening set, we do a similar reduction, using the actual scalars above.
-    let opening_set_reductions: Vec<Target> = proof.all_opening_sets().iter()
-        .map(|opening_set| reduce_with_coefficients(
-            builder, &opening_set.to_vec(), &actual_scalars))
+    let opening_set_reductions: Vec<Target> = proof
+        .all_opening_sets()
+        .iter()
+        .map(|opening_set| {
+            reduce_with_coefficients(builder, &opening_set.to_vec(), &actual_scalars)
+        })
         .collect();
 
     // Then, we reduce the above opening set reductions to a single value.
     let reduced_opening = reduce_with_powers_recursive(builder, &opening_set_reductions, v);
 
-    verify_ipa::<C, InnerC>(builder, proof, c_reduction, reduced_opening, x, ipa_challenges, security_bits)
+    let generator_n = builder.constant_wire(C::ScalarField::primitive_root_of_unity(
+        ipa_challenges.len(),
+    ));
+    let mut points: Vec<Target> = (0..2 * num_public_input_gates)
+        .step_by(2)
+        .map(|i| builder.exp_constant_usize(generator_n, i))
+        .collect();
+    points.extend(&[zeta, builder.mul(zeta, generator_n), {
+        let g65 = builder.exp_constant_usize(generator_n, GRID_WIDTH);
+        builder.mul(zeta, g65)
+    }]);
+    let halo_bs = points
+        .into_iter()
+        .map(|p| halo_g_recursive(builder, p, &ipa_challenges))
```

Comment:
In my mind `ipa_challenges` has the "raw" challenges (i.e. the randomness the verifier samples), but it's different from the `u` values `g` expects because of the endo. It might be a significant amount of work to fix all this stuff related to the IPA challenges, but could you add a TODO to note that this isn't the data `g` expects?

Reply:
Couldn't we store `n(u)` in `ipa_challenges` instead?

Reply:
By the way, I think this
```rust
halo_a = C::ScalarField::add_slices(
    &l_challenge_sf.scale_slice(a_lo),
    &r_challenge_sf.scale_slice(a_hi),
);
halo_b = C::ScalarField::add_slices(
    &l_challenge_sf.scale_slice(b_hi),
    &r_challenge_sf.scale_slice(b_lo),
);
```
in the prover (`plonk.rs` 371-378`) is wrong too, and actually all the proof/verification uses the "wrong" challenges. If we just apply `n` before storing in `ipa_challenges`, it should work though.